lua/bufferline/render.lua

```diff
@@ -757,125 +523,112 @@ local function generate_tabline(bufnrs, refocus)
 
-      bufferIndexPrefix = hl_tabline('Buffer' .. status .. 'Index')
-      bufferIndex = number_text .. ' '
-    end
+    --- The devicon
+    --- @type bufferline.render.group
+    local icon = {hl = clickable, text = ''}
 
     if state.is_picking_buffer then
       local letter = JumpMode.get_letter(bufnr)
 
       -- Replace first character of buf name with jump letter
-      if letter and not has_file_icons then
-        name = strcharpart(name, 1)
+      if letter and not icons_option.filetype.enabled then
+        name.text = strcharpart(name.text, 1)
       end
 
-      jumpLetterPrefix = hl_tabline('Buffer' .. status .. 'Target')
-      jumpLetter = (letter or '') ..
-        (has_file_icons and (' ' .. (letter and '' or ' ')) or '')
-    else
-
-      if has_file_icons then
-        local iconChar, iconHl = icons.get_icon(bufnr, status)
-        local hlName = is_inactive and not options.highlight_inactive_file_icons() and 'BufferInactive' or iconHl
-        iconPrefix = has_icon_custom_colors and hl_tabline('Buffer' .. status .. 'Icon') or (hlName and hl_tabline(hlName) or namePrefix)
-        icon = iconChar .. ' '
-      end
+      jump_letter.hl = hl_tabline('Buffer' .. activity .. 'Target')
+      jump_letter.text = (letter or '') ..
+        (icons_option.filetype.enabled and (' ' .. (letter and '' or ' ')) or '')
+    elseif icons_option.filetype.enabled then
+      local iconChar, iconHl = icons.get_icon(bufnr, activity)
+      local hlName = (activity == 'Inactive' and not options.highlight_inactive_file_icons())
+        and 'BufferInactive'
+        or iconHl
+
+      icon.hl = icons_option.filetype.custom_color and
+        hl_tabline('Buffer' .. activity .. 'Icon') or
+        (hlName and hl_tabline(hlName) or buffer_hl)
+      icon.text = iconChar .. ' '
     end
 
-    local closePrefix = ''
-    local close = ' '
-    if has_close or is_pinned then
-      local closeIcon =
-        is_pinned and
-          options.icon_pinned() or
-        (not is_modified and
-          options.icon_close_tab() or
-          options.icon_close_tab_modified())
-
-      closePrefix = namePrefix
-      close = closeIcon .. close
-
-      if click_enabled then
-        closePrefix =
-            '%' .. bufnr .. '@BufferlineCloseClickHandler@' .. closePrefix
-      end
-    end
+    --- The padding
+    --- @type bufferline.render.group
+    local padding = {hl = '', text = (' '):rep(layout.padding_width)}
 
-    local clickable = ''
-    if click_enabled then
-      clickable = '%' .. bufnr .. '@BufferlineMainClickHandler@'
-    end
-
-    local padding = (' '):rep(layout.padding_width)
+    --- The separator
+    --- @type bufferline.render.group
+    local separator = {
+      hl = clickable .. hl_tabline('Buffer' .. activity .. 'Sign'),
+      text = icons_option.separator,
+    }
 
     local item = {
+      groups = {separator, padding, buffer_index, buffer_number, icon, jump_letter, name},
```

Comment:
Here is the result of ([context](https://github.com/romgrk/barbar.nvim/pull/386/files#r1145488777)). You can visualize the left→right ordering of the buffer being rendered, and use `vim.lsp.buf.definition()` to modify what it is.